filters/include/pcl/filters/impl/farthest_point_sampling.hpp

```diff
@@ -0,0 +1,126 @@
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#pragma once
+
+#include <pcl/common/geometry.h>
+#include <pcl/filters/farthest_point_sampling.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <algorithm>
+#include <limits>
+#include <random>
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (PointCloud &output)
+{
+  Indices indices;
+  indices.reserve(sample_);
+  if (keep_organized_)
+  {
+    bool temp = extract_removed_indices_;
+    extract_removed_indices_ = true;
+    applyFilter (indices);
+    extract_removed_indices_ = temp;
+
+    output = *input_;
+    for (const auto ri : *removed_indices_)  // ri = removed index
+      output.points[ri].x = output.points[ri].y = output.points[ri].z = user_filter_value_;
+    if (!std::isfinite (user_filter_value_))
+      output.is_dense = false;
+  }
+  else
+  {
+    output.is_dense = true;
+    applyFilter (indices);
+    pcl::copyPointCloud (*input_, indices, output);
+  }
+}
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (std::vector<int> &indices)
+{
+  std::size_t size = input_->points.size();
+
+  //check if number of points in cloud is less than requested number of points
+  if (size <= sample_)
+  {
+    PCL_THROW_EXCEPTION (BadArgumentException,
+                        "Requested number of points is greater than point cloud size!");
+    return;
+  }
+
+  std::vector<float> distances_to_selected_points (size, std::numeric_limits<float>::max ());
+  
+  //set random seed
+  std::mt19937 random_gen(seed_);
+  std::uniform_int_distribution<> dis(0, size -1);
+
+  //pick the first point at random
+  int max_index = dis(random_gen);
+  distances_to_selected_points[max_index] = -1.0;
+  indices.push_back(max_index);
+
+  for (int j = 1; j < sample_; j++)
+  {
+    //recompute distances
+    for (int i = 0; i < distances_to_selected_points.size(); i++)
+    {
+      if (distances_to_selected_points[i] != -1.0)
+        distances_to_selected_points[i] = std::min(distances_to_selected_points[i], geometry::distance(input_->points[i], input_->points[max_index]));
+    }
+
+    //select farthest point based on previously calculated distances
+    //since distance is set to -1 for all selected elements,previously selected 
+    //elements are guaranteed to not be selected
+    max_index = std::distance(distances_to_selected_points.begin(), std::max_element(distances_to_selected_points.begin(), distances_to_selected_points.end()));
+    distances_to_selected_points[max_index] = -1.0;
+    indices.push_back(max_index);
+    //set distance to -1 to ignore during max element search
+  }
+
+  if (extract_removed_indices_)
+  {
+    for (int k = 0; k < distances_to_selected_points.size(); k++)
+    {
+      if (distances_to_selected_points[k] != -1.0)
+        (*removed_indices_).push_back(k);
+    }
+  }
+}
```

Comment:
Since you're making an extra pair of checks here as well, might was well populate `removed_indices_` inside the previous loop.

Reply:
Wouldn't `removed_indices_` be unknown till all `sample_` points have been sampled?

Reply:
I'm stupid :man_facepalming: